File: app/src/main/java/vocabletrainer/heinecke/aron/vocabletrainer/lib/ViewModel/SurveyViewModel.java
package vocabletrainer.heinecke.aron.vocabletrainer.lib.ViewModel;

import android.os.Build;
import android.util.Log;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.MalformedURLException;
import java.net.URL;

import javax.net.ssl.HttpsURLConnection;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import info.guardianproject.netcipher.NetCipher;

/**
 * ViewModel for survey, doing survey submit
 */
public class SurveyViewModel extends ViewModel {
    private static final String TAG = "SurveyViewModel";
    private MutableLiveData<Boolean> loading;
    private MutableLiveData<Boolean> error;
    private Thread runner;

    public SurveyViewModel() {
        loading = new MutableLiveData<>();
        error = new MutableLiveData<>();
        loading.setValue(false);
        error.setValue(false);
    }

    public synchronized void submitSurvey() {
        if (wasRunning()) {
            return;
        }
        loading.postValue(true);
        System.setProperty("http.keepAlive", "false");

        final String url;
        // allow tls 1.0
        // can't be tested fully due to different device libraries
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
            url = "https://vct.badssl.proctet.net/data/add/api";
        } else {
            url = "https://vct.proctet.net/data/add/api";
        }

        runner = new Thread(() -> {

            final JSONObject jsonBody;
            try {
                HttpsURLConnection con = NetCipher.getHttpsURLConnection(new URL(url));
                con.setRequestProperty("Content-Type", "application/json; charset=UTF-8");
                con.setRequestProperty("Accept", "application/json");
                con.setDoOutput(true);
                con.setDoInput(true);

                jsonBody = new JSONObject("{\"api\":" + Build.VERSION.SDK_INT + "}");

                OutputStream wr = con.getOutputStream();
                wr.write(jsonBody.toString().getBytes("UTF-8"));
                wr.flush();
                wr.close();

                InputStream in = new BufferedInputStream(con.getInputStream());
                ByteArrayOutputStream result = new ByteArrayOutputStream();
                byte[] buffer = new byte[1024];
                int length;
                while ((length = in.read(buffer)) != -1) {
                    result.write(buffer, 0, length);
                }

                JSONObject jsonObject = new JSONObject(result.toString("UTF-8"));
                in.close();

                con.disconnect();
                Log.d(TAG,"received json: "+jsonObject.toString());
                loading.postValue(false);
            } catch (
                    JSONException e) {
                Log.wtf(TAG, e);
                error.postValue(true);
            } catch (MalformedURLException e) {
                Log.wtf(TAG, e);
                error.postValue(true);
            } catch (IOException e) {
                Log.e(TAG, "IOException:",e);
                error.postValue(true);
            }
        });
        runner.start();
    }

    public boolean wasRunning() {
        return runner != null;
    }

    public LiveData<Boolean> getErrorLiveData() {
        return this.error;
    }

    public LiveData<Boolean> getLoadingLiveData() {
        return this.loading;
    }
}

File: app/src/main/java/vocabletrainer/heinecke/aron/vocabletrainer/lib/ViewModel/SurveyViewModel.kt
package vocabletrainer.heinecke.aron.vocabletrainer.lib.ViewModel

import org.acra.ACRA.errorReporter
import androidx.lifecycle.ViewModel
import androidx.lifecycle.MutableLiveData
import kotlin.jvm.Synchronized
import org.json.JSONObject
import android.os.Build
import android.util.Log
import androidx.lifecycle.LiveData
import org.acra.ktx.sendWithAcra
import vocabletrainer.heinecke.aron.vocabletrainer.BuildConfig
import java.io.BufferedInputStream
import java.io.ByteArrayOutputStream
import java.io.InputStream
import java.lang.Exception
import java.net.HttpURLConnection
import java.net.URL
import java.nio.charset.StandardCharsets

/**
 * ViewModel for survey, doing survey submit
 */
class SurveyViewModel : ViewModel() {
    private val loading: MutableLiveData<Boolean>
    private val error: MutableLiveData<Boolean>
    private var runner: Thread? = null
    @Synchronized
    fun submitSurvey() {
        if (wasRunning()) {
            return
        }
        loading.postValue(true)
        System.setProperty("http.keepAlive", "false")
        val url: String = if (BuildConfig.BUILD_TYPE == "debug") {
            //"http://192.168.178.32:3219/data/add/api"
            "http://localhost:3219/data/add/api"
        } else {
            // allow tls 1.0
            // can't be tested fully due to different device libraries
            "https://stats.vocabletrainer.de/data/add/api"
        }
        runner = Thread {
            val jsonBody: JSONObject
            try {
                errorReporter.putCustomData("url", url)
                val con = URL(url).openConnection() as HttpURLConnection
                con.setRequestProperty("Content-Type", "application/json; charset=UTF-8")
                con.setRequestProperty("Accept", "application/json")
                con.doOutput = true
                con.doInput = true
                jsonBody = JSONObject("{\"api\":" + Build.VERSION.SDK_INT + "}")
                val wr = con.outputStream
                wr.write(jsonBody.toString().toByteArray(StandardCharsets.UTF_8))
                wr.flush()
                wr.close()
                val `in`: InputStream = BufferedInputStream(con.inputStream)
                val result = ByteArrayOutputStream()
                val buffer = ByteArray(1024)
                var length: Int
                while (`in`.read(buffer).also { length = it } != -1) {
                    result.write(buffer, 0, length)
                }

                val resultString = result.toString("UTF-8")
                errorReporter.putCustomData("server response", resultString)
                val jsonObject = JSONObject(resultString)
                `in`.close()
                con.disconnect()
                Log.d(TAG, "received json: $jsonObject")
                loading.postValue(false)
            } catch (e: Exception) {
                Log.wtf(TAG, e)
                e.sendWithAcra()
                error.postValue(true)
            }
        }
        runner!!.start()
    }

    fun wasRunning(): Boolean {
        return runner != null
    }

    val errorLiveData: LiveData<Boolean>
        get() = error
    val loadingLiveData: LiveData<Boolean>
        get() = loading

    companion object {
        private const val TAG = "SurveyViewModel"
    }

    init {
        loading = MutableLiveData()
        error = MutableLiveData()
        loading.value = false
        error.value = false
    }
}